### DISCUSSION TITLE
Help with the right commands

### DISCUSSION THREAD

**Discussion by jhauseratuw** [TomWright/dasel#394] [Q&A] [answered]
First of all, wow is this fast!!!  THANK YOU!!  I am looking to take each array in the JSON example below and export it to .csv. e.g. Applications, Person, Interests, Majors, etc.  Using person.email as an added key to each .csv row. I have tried my best at reading the docs and comments. I am a bit stuck.

When using the command
 `dasel -r json 'all().ApplicationDetail' -w csv < /path/to/myfile.json` 
 The resulting .csv has the title row following each data row. e.g. 

row1: column1 column2
row2: field1, field2
row3: column1 column2
row4: field1, field2

What I am looking for of course is:
row1: column1 column2
row2: field1, field2
row3: field1, field2
etc.

When I try to move down to Interests for example: 
`dasel -r json 'all().ApplicationDetail.all()Interests' -w csv < /path/to/my/jsonfile.json`
I get a "Error: CSVParser.toBytes cannot handle type int64"

All help is appreciated!!  Here is one complete record of the JSON file for refernce

```
[
    {
        "ApplicationDetail": {
            "ApplicationID": 2000017729,
            "Email": "test@email.edu",
            "GPAComputed": 4.00,
            "GPARecent": 3.96,
            "Interests": [
                {
                    "InterestRank": 1,
                    "Interest": "interest 1"
                },
                {
                    "InterestRank": 2,
                    "Interest": "interest 2"
                }
            ],
            "Majors": [
                {
                    "DegreeIndex": 1,
                    "DegreeID": 109657
                }
            ],
            "Recommendations": [
                {
                    
                    "RecOrganization": "Univesity of Miami",
                    "RecPrefix": null,
                    "RecFirst": "Andrew"
                },
                {
                    "RecOrganization": "University of Missouri - Columbia",
                    "RecPrefix": null,
                    "RecFirst": "Joy"
                }
            ]
        },
        "Person": {
            "OfficialFirstName": "json",
            "OfficialLastName": "test",
            "Email": "test2@email.edu",
            "Contact": {
                "Phone": [
                    {
                        "PhoneRank": 1,
                        "PhoneNumber": "+1 555-555-5555",
                        "PhoneType": "Mobile"
                    }
                ],
                "Address": [
                    {
                        "AddressTypeName": "Permanent",
                        "AddressLine1": "1234 test street"
                    },
                    {
                        "AddressTypeName": "Local",
                        "AddressLine1": "3540 test lane"
                    }
                ]
            },
            "Education": [
                {
                    "InstitutionName": "University of Missouri-Columbia",
                    "AttendedFromYear": 2020
                }
            ],
            "Employment": [
                {
                    "EmployerName": "Ziply Fiber",
                    "EmployeeTitle": "Support Leader"
                },
                {
                    "EmployerName": "McDonalds",
                    "EmployeeTitle": "Intern"
                }
            ],
            "Tests": []
        }
    }
]`

**Comment by TomWright** [comment]
Part of the issue here is that you are using `ApplicationDetail.all().Interests`.

If we just do `ApplicationDetail.all()` you can see it's opening up a path down each field, some of which aren't objects.
```
$ dasel -f file.json 'all().ApplicationDetail.all()'
2000017729
"test@email.edu"
4
3.96
[
  {
    "InterestRank": 1,
    "Interest": "interest 1"
  },
  {
    "InterestRank": 2,
    "Interest": "interest 2"
  }
]
[
  {
    "DegreeIndex": 1,
    "DegreeID": 109657
  }
]
[
  {
    "RecOrganization": "Univesity of Miami",
    "RecPrefix": null,
    "RecFirst": "Andrew"
  },
  {
    "RecOrganization": "University of Missouri - Columbia",
    "RecPrefix": null,
    "RecFirst": "Joy"
  }
]
```

If we adjust the selector we can get closer:
```
$ dasel -f file.json -w csv 'all().ApplicationDetail.Interests.all().mapOf(PersonEmail,parent().parent().parent().Person.Email,InterestRank,InterestRank,Interest,Interest)'
Interest,InterestRank,PersonEmail
interest 1,1,test2@email.edu
Interest,InterestRank,PersonEmail
interest 2,2,test2@email.edu
```

Could I have an example CSV output for the given file above so I can give some more accurate help?

**Reply by jhauseratuw** [reply]
This answered many questions and is greatly appreciated. :) 
The csv output I am looking for is:

//For Interest
Interest,InterestRank,PersonEmail
interest 1,1,test@email.edu
interest 2,2,test@email.edu

//For ApplicationDetail
ApplicationID,Email,GPAComputed,GPARecent
2000017729,test@email.edu,4.00,3.96
(the next ApplicationDetail record without a header row which is beyond the scope of the data above e.g...)
2000017730,person2@email.edu,3.00,2.96

Related to this when I try to modify your working code for ApplicationDetail with a -w csv e.g.
dasel -f file.json -w csv 'all().ApplicationDetail.all()'
I get an error: Error: CSVParser.toBytes cannot handle type int64

Apologies in advance for my ignorance!

**Reply by jhauseratuw** [reply]
Upon further review, the following will remove the extra header rows for me. So scratch that request. :). What is left then is getting AppplicationDetail sorted. 

`dasel -f file.json -w csv 'all().ApplicationDetail.Interests.all().mapOf(PersonEmail,parent().parent().parent().Person.Email,InterestRank,InterestRank,Interest,Interest)' | awk 'NR==1 || NR%2==0' > file.csv`

**Reply by jhauseratuw** [reply, marked as answer]
Finally, I got it. Thank you for your patience, and quick response. You did indeed answer everything I needed. To get ApplicationDetial all I needed was:  
`dasel -f $json -w csv 'all().ApplicationDetail.mapOf(ApplicationID,ApplicationID)' | awk 'NR==1 || NR%2==0'`

You sir ar the BEST!!

**Reply by TomWright** [reply]
Hey @jhauseratuw , sorry for the delayed reply.

You can also use the following selector to get the output you want:

```
all().ApplicationDetail.Interests.all().mapOf(PersonEmail,parent().parent().parent().Person.Email,InterestRank,InterestRank,Interest,Interest).merge()
```

It's doing the same as before but merging the  arrays to avoid multiple heading lines.
